src/compact_live.rs

```diff
@@ -0,0 +1,454 @@
+            .await?;
+
+        // Fetch chain data (noisy).
+        let chains = self
+            .fetch_chain_data(
+                k_value,
+                &noisy_encrypted_entry_table,
+            )
+            .await?;
+
+        // Select remaining locations (noisy).
+        let noisy_locations: HashSet<Location> = chains.chain_values
+            .iter()
+            .flat_map(|(_, values)| values)
+            .filter_map(|value| value.get_location())
+            .cloned()
+            .collect();
+        let noisy_remaining_locations = self.filter_removed_locations(&noisy_locations)?;
+
+        // While there are some Entry Table elements to compact...
+        while !noisy_encrypted_entry_table.is_empty() {
+            // Compute the new Entry Table (noisy) and compact chains (not noisy).
+            let (noisy_entry_table, new_chains) = self.compact_chains(
+                rng,
+                k_value,
+                noise,
+                &noisy_remaining_locations,
+                &noisy_encrypted_entry_table,
+                &chains.chain_values,
+            )?;
+
+            // Insert all recompacted chains first.
+            self.insert_chain_table(&new_chains.iter().flat_map(|(_, v)| v.clone()).collect()).await?;
+
+            // Try upserting the new entry table.
+            let upsert_data = UpsertData::new(
+                &noisy_encrypted_entry_table,
+                noisy_entry_table.encrypt::<DEM_KEY_LENGTH, DemScheme>(k_value, rng)?,
+            );
+
+            // These are failures to upsert.
+            noisy_encrypted_entry_table = self.upsert_entry_table(&upsert_data).await?;
+
+            // Delete unused chains (at least one chain element per upsert try):
+            let mut chains_to_delete: HashSet<Uid<UID_LENGTH>> =
+                HashSet::with_capacity(upsert_data.len());
+            chains_to_delete.extend(
+                // - new chains corresponding to unsuccessful upserts
+                new_chains
+                    .iter()
+                    .filter_map(|(uid, chains)| {
+                        if noisy_encrypted_entry_table.contains_key(uid) {
+                            Some(chains.keys())
+                        } else {
+                            None
+                        }
+                    })
+                    .flatten()
+                    .cloned(),
+            );
+
+            chains_to_delete.extend(
+                // - old chains corresponding to successful upserts
+                chains.chain_uids
+                    .iter()
+                    .filter_map(|(uid, chain)| {
+                        if !noisy_encrypted_entry_table.contains_key(uid) && !noise.contains(uid) {
+                            Some(chain)
+                        } else {
+                            None
+                        }
+                    })
+                    .flatten()
+                    .cloned(),
+            );
```

Comment:
We may be able to do the same here (the for loop should be smaller) but the `.clone()` is required since `chains` is re-used multiple times inside the while loop.

Reply:
done in 9d26714